nucypher/blockchain/eth/currency.py

```diff
@@ -0,0 +1,197 @@
+from _pydecimal import Decimal
+
+import eth_utils
+import maya
+from eth_utils import currency
+from nacl.hash import sha256
+from typing import Union, Tuple
+
+from nucypher.blockchain.eth.agents import NucypherTokenAgent
+from nucypher.blockchain.eth.constants import SUBDIGITS
+from nucypher.blockchain.eth.utils import datetime_at_period, datetime_to_period
+
+
+class NU:
+    """
+    An amount of NuCypher tokens that doesn't hurt your eyes.
+    Wraps the eth_utils currency conversion methods.
+
+    The easiest way to use NU, is to pass an int, float, or str, and denomination string:
+
+    Int:    nu = NU(100, 'NU')
+    Int:    nu_wei = NU(15000000000000000000000, 'NUWei')
+
+    Float:  nu = NU(15042.445, 'NU')
+    String: nu = NU('10002.302', 'NU')
+
+    ...or alternately...
+
+    Float: nu = NU.from_tokens(100.50)
+    Int: nu_wei = NU.from_nu_wei(15000000000000000000000)
+
+    Token quantity is stored internally as an int in the smallest denomination,
+    and all arithmetic operations use this value.
+    """
+
+    __symbol = 'NU'
+    __decimals = SUBDIGITS
+    __agent_class = NucypherTokenAgent
+
+    # conversions
+    __denominations = {'NUWei': 'wei',
+                       'NU': 'ether'}
+
+    def __init__(self, value: Union[int, float, str], denomination: str):
+
+        # Calculate smallest denomination and store it
+        wrapped_denom = self.__denominations[denomination]
+
+        # Validate Early
+        if '.' in str(value):
```

Comment:
I think we can write an issue and move on wrt to this particular problem, BTW.

Reply:
This boils down to not being able to properly validate a `NU` when instantiated with scientific notation.

For example  `NU(1e+99, 'NU')` will result in:
 `ValueError: Resulting wei value must be between 1 and 2**256 - 1` 

instead of:
 `ValueError("Cannot initialize with fractional wei value")`.

Reply:
This approach seems like a clear improvement:

```
decimal_part = Decimal(value)%1
if not decimal_part.is_zero():
    # The number had decimals
    ...
```

Reply:
https://github.com/nucypher/nucypher/issues/857

Reply:
@cygnusv - :thinking: - Actually, why *wont* this work, since `str(value)` will always add the decimal?  

For example:

```
str(1e+12)
'1000000000000.0'
```